# One Hot Encoding using pandas and numpy

### It is an essential process of converting the categorical data variables to be provided to machine and deep learning algorithms which in turn improve predictions as well as classification accuracy of a model

In [149]:
import pandas as pd
import numpy as np

In [150]:
data=pd.read_csv("benz.csv",usecols=["X2","X3","X5","X1"])
data.head()

X1  X2 X3 X5
0  v  at  a  u
1  t  av  e  y
2  w   n  c  x
3  t   n  f  x
4  v   n  f  h

In [151]:
for column in data.columns:
    print(column,len(data[column].unique()))

X1 27
X2 44
X3 7
X5 29


In [152]:
pd.get_dummies(data).shape

(4209, 107)

We can see 4 columns are encoded to 107 columns.
What if we have many unique values in each column and no of columns are n.
Then no of columns going to be very large whcih leads to increase in dimensionality of dataset and accuracy of model decreases.
To avoid this, we can seclect most frequent top m values in each column.

In [153]:
# Getting all unique values in column and total no of apperances in each column
for column in data.columns:
    print(data[column].value_counts().sort_values(ascending=False))

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64
as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
d       18
ac      13
g       12
ap      11
y       11
x       10
aw       8
at       6
h        6
al       5
an       5
q        5
p        4
ah       4
av       4
au       3
am       1
j        1
af       1
l        1
aa       1
c        1
o        1
ar       1
Name: X2, dtype: int64
c    1942
f    1076
a     440
d     290
g     241
e     163
b      57
Name: X3, dtype: int64
w     231
v     231
q     220
r     215
s     214
d     214
n     212
p     208
m     208
i     207
ae    2

In [154]:
# Select top m values that are most freqent in each column
# Getting top m values for each column in dataset
Top10Values=[]
m=10
for column in data.columns:
    Top10Values.append(data[column].value_counts().sort_values(ascending=False).head(m).index)
    
# Top10Values=pd.DataFrame(Top10Values)
Top10Values

[Index(['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o'], dtype='object'),
 Index(['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e'], dtype='object'),
 Index(['c', 'f', 'a', 'd', 'g', 'e', 'b'], dtype='object'),
 Index(['w', 'v', 'q', 'r', 's', 'd', 'n', 'p', 'm', 'i'], dtype='object')]

In [155]:
cols=data.columns
for i in range(data.shape[1]):
    # without using dummy variable trap
    for label in Top10Values[i]:
        if(label!=None):
            data[cols[i]+"_"+label]=np.where(data[cols[i]]==label,1,0)

In [156]:
data.drop(columns=cols,inplace=True)
data.shape

(4209, 37)

In [157]:
data

X1_aa  X1_s  X1_b  X1_l  X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  ...  X5_w  \
0         0     0     0     0     1     0     0     0     0     0  ...     0   
1         0     0     0     0     0     0     0     0     0     0  ...     0   
2         0     0     0     0     0     0     0     0     0     0  ...     0   
3         0     0     0     0     0     0     0     0     0     0  ...     0   
4         0     0     0     0     1     0     0     0     0     0  ...     0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4204      0     1     0     0     0     0     0     0     0     0  ...     0   
4205      0     0     0     0     0     0     0     0     0     1  ...     0   
4206      0     0     0     0     1     0     0     0     0     0  ...     0   
4207      0     0     0     0     0     1     0     0     0     0  ...     0   
4208      0     0     0     0     0     1     0     0     0     0  ...     0   

      X5_v  X5_q  X5_r  X5_s  X5_d  X5_n  X5_p  X5_m  X5_i  
0        0     0     0     0     0     0     0     0     0  
1        0     0     0     0     0     0     0     0     0  
2        0     0     0     0     0     0     0     0     0  
3        0     0     0     0     0     0     0     0     0  
4        0     0     0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0     0     0     0  
4205     0     0     0     0     0     0     0     0     0  
4206     0     0     0     0     0     0     0     0     0  
4207     0     0     0     0     0     0     0     0     0  
4208     0     0     0     0     0     0     0     0     0  

[4209 rows x 37 columns]